rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,260 @@
+# RFC 8216 - 2021-07-19 - multiple pipelines
+
+Large Vector users often require complex Vector topologies to facilitate the collection and processing of data from many different upstream sources. This results in large Vector configuration files that are hard to manage across different teams. This makes Vector a bad candidate for large teams that require autonomy to facilitate sane collaboration.
+
+## Scope
+
+This RFC will cover:
+
+- How and where the pipelines are stored
+- How vector reads the pipelines and what are the limits of pipelines
+- What metadatas vector adds to the events going through a pipeline
+
+This RFC will not cover:
+
+- How the pipelines should be synchronized between vector instances
+- Pre/post processing of data - The ability to prepare and normalize data.
+- Connecting pipelines together - The ability to take input from another pipeline.
+- Component reuse - The ability to define boilerplate for reuse across many different pipelines. This will likely align with Datadog’s “pipeline catalogue”.
+- Pipeline quotas - The ability to limit how much data a pipeline can send to a sink.
+
+
+## Motivation
+
+- Helps Vector to grow organically within an organization by allowing teams to adopt Vector at their own pace without heavy administrator involvement.
+- Reduces the management overhead of devops/SREs by enabling teams to manage their own pipelines (spread the management load).
+
+
+## Internal Proposal
+
+### How and where the pipelines are stored
+
+To keep the actual structure and avoid retro compatibility issues, vector should read from the default folder `/etc/vector/pipelines/*` the difference pipeline's configuration files.
+The pipeline directory should have a flat structure (no sub directories) and only `json`, `toml` and `yaml` files will be supported.
+
+To discuss: make this folder should be configurable through the cli, the same way than for the configuration file, with `--pipelines-dir` or though and environment variable `VECTOR_PIPELINES_DIR`.
+
+### How vector reads the pipelines and what are the limits
+
+Vector will read each of the pipeline configuration files at start time and create an `id` attribute corresponding to the pipeline's filename. A `load-balancer.yml` will have `load-balancer` as `id`. In the pipeline's configuration file, the `id` can be set manually.
```

Comment:
> at start time

With regard to vector's boot, today we support live reloads of configuration. I've got appetite to drop this feature entirely but it's not something we've talked about broadly in the team, so please do consider whether pipelines are a reload capable feature or if we should fast-track the conversation around dropping live reload.

Reply:
I noticed this live reload feature yesterday and didn't think about changing this. 👍